saleor/discount/utils/voucher.py

```diff
@@ -88,7 +88,7 @@ def increase_voucher_code_usage_value(code: "VoucherCode") -> None:
 
 def decrease_voucher_code_usage_value(code: "VoucherCode") -> None:
     """Decrease voucher code uses by 1."""
-    code.used = F("used") - 1
+    code.used = F("used") - 1 if code.used >= 1 else 0
```

Comment:
This would run an extra save in case of `0`. Also, I think it could be prone to race-condition errors when two workers run the function at the same time, both see `code.used = 1` and then both subtract 1 (but it's just a theory). We could make it work entirely at the DB level, which should prevent both issues:

```
VoucherCode.objects.filter(pk=code.pk).update(
        used=Case(
            When(used__gt=0, then=F("used") - 1),
            default=Value(0),
            output_field=IntegerField(),
        )
    )
```

Reply:
Done